### PR TITLE
ensure that requests timeout even when initOp

### PR DESCRIPTION
**Pull request by Raynos**
This improves the pre-flight waitForIdentified() logic
in the tchannel.Request class.

Here we will actually do a check to see if we have
a usable connection and skip the waiting.

If we have to wait because there are zero connections
we also allocate a RequestOperation with a properly
configured timeout and add it to the timeHeap.

This means that the request will timeout within it's
specified `timeout` irregardless of what the connection
initTimeout value is.

We clean up and cancel the timeout once a connection
is identified and available

r: @zhijin @syyang

### CODE REVIEW

**Comment by zhijinli** [comment]
LGTM

File: request.js [hunk 1]
```diff
@@ -80,6 +80,23 @@ inherits(TChannelRequest, EventEmitter);
...
+    this.timeout = timeout;
+    this.peer = peer;
+    this.waitForIdentifiedSlot = waitForIdentifiedSlot;
+}
+
+RequestOperation.prototype.onTimeout = function onTimeout(now) {
+    if (this.waitForIdentifiedSlot !== null &&
+        this.waitForIdentifiedSlot !== -1
+    ) {
+        this.peer.stopWaitingForIdentified(this.waitForIdentifiedSlot);
```
Comment (syyang) [inline request.js]: does `stopWaitingForIdentified` not do the null or -1 check?

File: request.js [hunk 2]
```diff
@@ -80,6 +80,23 @@ inherits(TChannelRequest, EventEmitter);
...
+function RequestOperation(req, timeout, peer, waitForIdentifiedSlot) {
+    this.req = req;
+    this.timeout = timeout;
+    this.peer = peer;
+    this.waitForIdentifiedSlot = waitForIdentifiedSlot;
+}
+
+RequestOperation.prototype.onTimeout = function onTimeout(now) {
+    if (this.waitForIdentifiedSlot !== null &&
+        this.waitForIdentifiedSlot !== -1
```
Comment (syyang) [inline request.js]: maybe default `waitForIdentifiedSlot` to `-1` above?

File: request.js [hunk 3]
```diff
@@ -304,9 +321,25 @@ TChannelRequest.prototype.resend = function resend() {
...
+        self.elapsed = now - self.start;
+        var timeout = self.timeout - self.elapsed;
+        var reqOp = new RequestOperation(
+            self, timeout, peer, waitForIdentifiedSlot
+        );
+        reqTimeout = self.channel.timeHeap.update(reqOp, now);
+    }
 
     function onIdentified(err) {
+        reqTimeout.cancel();
```
Comment (syyang) [inline request.js]: some comment on why cancel needs to be called will be useful. isn't `reqTimeout` undefined at this point?

File: request.js (hunk 2 above)
Comment (Raynos) [inline request.js]: Sure.

**Comment by syyang** [comment]
lgtm. thanks!

**Pull request merged**